### PR TITLE
semgrepReport (JSON) import Scans

### PR DESCRIPTION
**Pull request by SPoint42**
Supercedes #2777

This template is for your information. Please clear everything when submitting your pull request.

**Note: DefectDojo is now on Python3.6 and Django 2.2.x. Please submit your pull requests to the 'dev' branch as the 'legacy-python2.7' branch is only for bug fixes. Any new features submitted to the legacy branch will be ignored and closed.**

When submitting a pull request, please make sure you have completed the following checklist:

- [x] Give a meaningful name to your PR, as it may end up being used in the release notes.
- [x] Your code is flake8 compliant.
- [x] Your code is python 3.6 compliant (specific python >3.6 syntax is currently not accepted).
- [ ] If this is a new feature and not a bug fix, you've included the proper documentation in the ReadTheDocs documentation folder. https://github.com/DefectDojo/Documentation/tree/master/docs or provide feature documentation in the PR.
- [ ] Model changes must include the necessary migrations in the dojo/db_migrations folder.
- [x] Add applicable tests to the unit tests.
- [x] Add the proper label to categorize your PR.

### CODE REVIEW

**Comment by SPoint42** [comment]
Just finish the reworking on the PR, the PR could be review, as it working against the provide example on semgrep file

File: dojo/tools/semgrep/models.py
Comment (DrewDennison) [inline dojo/tools/semgrep/models.py:37]: hey I'm one of the maintainers of semgrep and I'm excited that DefectDojo is integrating semgrep. Happy to help if needed with things. I noticed that this line will error out if `metadata.cwe` is not defined and will throw 

```
AttributeError: 'NoneType' object has no attribute 'partition'
```

perhaps change to `metadata.get("cwe", "")`?

**Comment by xee5ch** [comment]
@SPoint42, I guess this is my fault for not pushing up my changes and tests on Sunday afternoon, as you have now pushed ahead of me. Let me know if you would like to carry it forward and I back off, up to you.

I see you made a separate branch so either way!

File: dojo/tools/semgrep/models.py
Comment (xee5ch) [reply dojo/tools/semgrep/models.py:37]: Yeah, I guess that makes sense. I was telling Sebastien he has adapted some of my code in this PR off of a fork of my fork it seems, and I just saw the emails now. I was not splitting on colon at all, so I did not even notice, and I was not sure yet from playing which fields in the `"metadata": {}` struct are optional and mandatory from looking at semgrep code (I am OK with Python, less so with OCaml).

But I defer to the majority opinion here. I worked on this to force myself to start evaluating semgrep more seriously, where as Sebastian is actually using it and wants it in DD, so his opinions are probably way more informed than mine!

Comment (SPoint42) [reply dojo/tools/semgrep/models.py:37]: Will make a patch tomorrow. This seems to be very simple to handle @DrewDennison comment

**Comment by SPoint42** [comment]
> @SPoint42, I guess this is my fault for not pushing up my changes and tests on Sunday afternoon, as you have now pushed ahead of me. Let me know if you would like to carry it forward and I back off, up to you.
> 
> I see you made a separate branch so either way!

Up to you, you can pull it on my fork or let me continue ?

**Comment by xee5ch** [comment]
> > @SPoint42, I guess this is my fault for not pushing up my changes and tests on Sunday afternoon, as you have now pushed ahead of me. Let me know if you would like to carry it forward and I back off, up to you.
> > I see you made a separate branch so either way!
> 
> Up to you, you can pull it on my fork or let me continue ?

I pushed up my code for you to take a look, it includes tests, but I need to look at your code and how it compares it. It is up to you!

**Comment by SPoint42** [comment]
> > > @SPoint42, I guess this is my fault for not pushing up my changes and tests on Sunday afternoon, as you have now pushed ahead of me. Let me know if you would like to carry it forward and I back off, up to you.
> > > I see you made a separate branch so either way!
> > 
> > 
> > Up to you, you can pull it on my fork or let me continue ?
> 
> I pushed up my code for you to take a look, it includes tests, but I need to look at your code and how it compares it. It is up to you!

I could have time to review it if you want, let me know. If yes, send me in private

**Comment by Maffooch** [comment]
Hey guys, I noticed this has been converted from a draft to a regular PR. Is this the one y'all are going with instead of #2767?

**Comment by xee5ch** [comment]
> Hey guys, I noticed this has been converted from a draft to a regular PR. Is this the one y'all are going with instead of #2767?

We working in parallel by accident and now trying to work that out. I guess we will let you know ... soon?

**Comment by SPoint42** [comment]
> Hey guys, I noticed this has been converted from a draft to a regular PR. Is this the one y'all are going with instead of #2767?

Yes, PR2767 was not working, and now this is a working version. Sorry for this

File: dojo/tools/semgrep/models.py
Comment (SPoint42) [reply dojo/tools/semgrep/models.py]: done

**Comment by SPoint42** [comment]
> @SPoint42 you need help with tests and stuff?

did see it, and think will not have time to. Is it really necessary ?

**Comment by Maffooch** [comment]
We would really appreciate tests. Check out #2612 for what to go for.

**Comment by Maffooch** [comment]
Thanks for adding the tests! Please fix commits and I'll approve.

File: dojo/forms.py
Comment (Maffooch) [inline dojo/forms.py:405]: Was this meant to be removed?

Comment (SPoint42) [reply dojo/forms.py:405]: I think this was a mistake
DrHeader is already imported

File: dojo/fixtures/test_type.json
Comment (valentijnscholten) [inline dojo/fixtures/test_type.json:655]: looks like you're overwriting drheader importer?

File: dojo/forms.py
Comment (valentijnscholten) [reply dojo/forms.py:416]: looks ok here

**Review by valentijnscholten — CHANGES_REQUESTED** [review]

**Comment by madchap** [comment]
@SPoint42 Any updates per the review? I was looking into that today, since I was playing with semgrep, thought it had been merged already :-) Cheers.

File: dojo/fixtures/test_type.json
Comment (SPoint42) [reply dojo/fixtures/test_type.json:655]: Yes, was a bad commit. 
We do not need to override it, but add a new PK assign to Semgrep.

**Comment by madchap** [comment]
@SPoint42 please review the integration tests and flake8 results. It has to be all green. Thanks.

File: dojo/tools/semgrep/parser.py
Comment (madchap) [inline dojo/tools/semgrep/parser.py:31]: `active` and `verified` can be removed, this will be anyways be overridden.

File: dojo/tools/semgrep/parser.py
Comment (madchap) [inline dojo/tools/semgrep/parser.py:33]: No need to specify `N/A` if nothing to input. 
However, could `self.source_rule_url` be used to fill in that field?

File: dojo/tools/semgrep/models.py
Comment (madchap) [inline dojo/tools/semgrep/models.py:41]: Is this used anywhere? Could it be appended to `references`?

File: dojo/tools/semgrep/parser.py
Comment (madchap) [inline dojo/tools/semgrep/parser.py:26]: In my test file, I have a message that contains quite a few `\n`, which oddly breaks the description in defectdojo. Do you think it'd be a good thing to ensure newline characters are removed before being inserted into DefectDojo?

My example
```
Found object deserialization using ObjectInputStream. Deserializing entire\nJava objects is dangerous because malicious actors can create Java object\nstreams with unintended consequences. Ensure that the objects being deserialized\nare not user-controlled. If this must be done, consider using HMACs to sign\nthe data stream to make sure it is not tampered with, or consider only\ntransmitting object fields and populating a new object.\n"
```

**Review by madchap — CHANGES_REQUESTED** [review]
Added a few points. Overall, parser is working, I just would like to eventually see more of semgrep data flowing in. Cheers!

File: dojo/tools/semgrep/parser.py
Comment (SPoint42) [reply dojo/tools/semgrep/parser.py:26]: Really don"t know. I did not have the case in my whole tests .

File: dojo/tools/semgrep/parser.py
Comment (SPoint42) [reply dojo/tools/semgrep/parser.py:33]: not sure, because in my understanding of DojoFindings, url is related to a test url ? did I miss something ?

File: dojo/tools/semgrep/models.py
Comment (SPoint42) [reply dojo/tools/semgrep/models.py:41]: yes seem better

File: dojo/templates/dojo/import_scan_results.html
Comment (valentijnscholten) [inline dojo/templates/dojo/import_scan_results.html:83]: please stick to alphabetical ordering

File: dojo/tools/semgrep/parser.py
Comment (valentijnscholten) [inline dojo/tools/semgrep/parser.py:20]: could you use a more descriptive variable name, i.e. semgrep_result instead of f?

**Review by valentijnscholten — CHANGES_REQUESTED** [review]
could you double check everywhere where you reference items/values/attributes from the report they are either always there or the parser has a check to preven it from referencing none existing items which would raise a 500 error?

File: dojo/tools/semgrep/parser.py
Comment (madchap) [reply dojo/tools/semgrep/parser.py:31]: Last update to what is was is "pointing to an external source to explain the finding". Otherwise, please remove altogether.

**Review by madchap — APPROVED** [review]

**Review by valentijnscholten — APPROVED** [review]

**Comment by oksid** [comment]
Hey!) import semgrep scans in defect dojo work?

**Comment by damiencarol** [comment]
@oksid with this pull request, the import of JSON report works now. so I think it will make it for version `1.13.0`

**Pull request merged**
